### PR TITLE
Add GridSample Operator

### PR DESCRIPTION
**Pull request by hwangdeyu**
This PR proposes a new operator `Grid_sample`.
This op is often used for [Spatial Transformer Networks](https://arxiv.org/abs/1506.02025).
See the op's doc for more details.

### CODE REVIEW

File: docs/Changelog.md [hunk 1]
```diff
@@ -18939,6 +18939,49 @@ This version of the operator has been available since version 14 of the default
...
+
+#### Attributes
+
+<dl>
+<dt><tt>align_corners</tt> : int (default is 0)</dt>
+<dd>If align_corners=1, the extrema (-1 and 1) are considered as referring to the center points of the input's corner pixels. If align_corners=0, they are instead considered as referring to the corner points of the input's corner pixels, making the sampling more resolution agnostic.</dd>
+<dt><tt>mode</tt> : string (default is bilinear)</dt>
+<dd>Three interpolation modes: bilinear (default), nearest and bicubic.</dd>
+<dt><tt>padding_mode</tt> : string (default is zeros)</dt>
+<dd>Supported Padding modes for outside grid values: `zeros`(default), `border`, `reflection`</dd>
```
Comment (askhade) [inline docs/Changelog.md:18959]: add explaination for different padding modes -

File: docs/Changelog.md [hunk 2]
```diff
@@ -18939,6 +18939,49 @@ This version of the operator has been available since version 14 of the default
...
+
+#### Version
+
+This version of the operator has been available since version 14 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>align_corners</tt> : int (default is 0)</dt>
+<dd>If align_corners=1, the extrema (-1 and 1) are considered as referring to the center points of the input's corner pixels. If align_corners=0, they are instead considered as referring to the corner points of the input's corner pixels, making the sampling more resolution agnostic.</dd>
```
Comment (askhade) [inline docs/Changelog.md:18955]: the align_corners option selected here should be in sync\same as with the option selected for resizing the image (if necessary)  before grid sampling right?

File: docs/Changelog.md [hunk 3]
```diff
@@ -18939,6 +18939,49 @@ This version of the operator has been available since version 14 of the default
...
+<dd>Supported Padding modes for outside grid values: `zeros`(default), `border`, `reflection`</dd>
+</dl>
+
+#### Inputs
+
+<dl>
+<dt><tt>X</tt> (differentiable) : T</dt>
+<dd>4-D tensor of shape (N, C, inH, inW), where N is the batch size, C is the numbers of channels, inH and inW are the height and width of the data.</dd>
+<dt><tt>grid</tt> (non-differentiable) : T</dt>
+<dd>4-D tensor of shape (N, outH, outW, 2), where outH and outW is the height and width of offset and output.</dd>
```
Comment (askhade) [inline docs/Changelog.md:18968]: Can you add details around possible values in grid, and how should the out of range values be treated

File: docs/TestCoverage.md [hunk 4]
```diff
@@ -14673,6 +14673,9 @@ expect(node, inputs=[x, y], outputs=[z],
 ### GreaterOrEqual (call for test cases)
 
 
+### GridSampler (call for test cases)
```
Comment (askhade) [inline docs/TestCoverage.md:14676]: need to add tests before finalizing the spec

File: onnx/defs/operator_sets.h [hunk 5]
```diff
@@ -942,6 +942,7 @@ class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 14, Sub);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 14, Mul);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 14, Div);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 14, Identity);
+class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 14, GridSampler);
```
Comment (askhade) [inline onnx/defs/operator_sets.h:945]: you should add this to opset 15 as this is the latest version

File: onnx/defs/tensor/defs.cc [hunk 6]
```diff
@@ -2226,6 +2226,65 @@ ONNX_OPERATOR_SET_SCHEMA(
           resizeShapeInference(ctx, true);
         }));
 
+static const char* GridSampler_ver13_doc = R"DOC(
```
Comment (askhade) [inline onnx/defs/tensor/defs.cc:2229]: nit: GridSampler_ver15_doc

File: onnx/defs/tensor/defs.cc [hunk 7]
```diff
@@ -2226,6 +2226,65 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
+        .TypeConstraint(
+            "T",
+            OpSchema::all_tensor_types(),
+            "Constrain input 'X', 'grid' and output 'Y' to all tensor types.")
+        .SetDoc(GridSampler_ver13_doc)
+        .TypeAndShapeInferenceFunction(propagateShapeAndTypeFromFirstInput));
```
Comment (askhade) [inline onnx/defs/tensor/defs.cc:2286]: propagateShapeAndTypeFromFirstInput will copy the shape from input to output... for this op this is not correct... you need to construct the shape from input and grid shape

File: onnx/defs/operator_sets.h (hunk 5 above)
Comment (hwangdeyu) [reply onnx/defs/operator_sets.h:945]: Now exporter only support opset 14 after this PR merged. https://github.com/pytorch/pytorch/pull/59486.
So I'm not sure if it's better to support exporter to community or 1-p customer in opset14, or it's needed to use the latest version for ONNX？

File: onnx/defs/tensor/defs.cc (hunk 7 above)
Comment (hwangdeyu) [reply onnx/defs/tensor/defs.cc:2286]: Thanks, that's the part I am doing..

File: onnx/defs/operator_sets.h (hunk 5 above)
Comment (askhade) [reply onnx/defs/operator_sets.h:945]: opset 14 is already released and you cannot add a new op to a released opset. Work for opset15 is in progress now and any op update or new op should be added to opset 15. Once onnx releases (next one is scheduled for July end) exporter can add support for this new opset.

File: onnx/defs/tensor/defs.cc (hunk 7 above)
Comment (gramalingam) [reply onnx/defs/tensor/defs.cc:2286]: Please take a look at the example here: https://github.com/onnx/onnx/blob/ea8e57a71e01d1e70b188465c028eb3c6c6cc118/onnx/defs/object_detection/defs.cc#L114 ... ```unifyInputDim``` should help simplify this.

File: docs/Changelog.md [hunk 8]
```diff
@@ -18939,6 +18939,49 @@ This version of the operator has been available since version 14 of the default
 <dd>Constrain seq_lens to integer tensor.</dd>
 </dl>
 
+### <a name="GridSampler-14"></a>**GridSampler-14**</a>
+
+  Given an input and a flow-field grid, computes the output using input values and pixel locations from grid.
```
Comment (hwangdeyu) [inline docs/Changelog.md:18944]: Consider if we also need to support volumetric (5-D) input support.

File: onnx/defs/tensor/defs.cc (hunk 6 above)
Comment (gramalingam) [reply onnx/defs/tensor/defs.cc:2229]: It is unclear to me what this function is supposed to compute. Can we expand the description to be more complete, or add a link to some existing function in other frameworks, or both? Thanks.

File: onnx/test/shape_inference_test.py [hunk 9]
```diff
@@ -3587,6 +3587,15 @@ def test_nonzero_no_shape(self):  # type: () -> None
             [])
         self._assert_inferred(graph, [make_tensor_value_info('out', TensorProto.INT64, (None, None))])  # type: ignore
 
+    def test_gridsampler(self):  # type: () -> None
```
Comment (hwangdeyu) [inline onnx/test/shape_inference_test.py:3590]: Add more shape inference test for different shape and param.

File: onnx/defs/tensor/defs.cc [hunk 10]
```diff
@@ -2226,6 +2226,65 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            "4-D tensor of shape (N, C, inH, inW), where N is the batch size, C is the numbers of channels, inH and inW are the height and width of the data.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
+        .Input(
+            1,
+            "grid",
+            "4-D tensor of shape (N, outH, outW, 2), where outH and outW is the height and width of offset and output.",
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:2266]: Is the "offset and output" a typo? Don't understand it.

File: onnx/defs/tensor/defs.cc (hunk 6 above)
Comment (askhade) [reply onnx/defs/tensor/defs.cc:2229]: I prefer both as opposed to just adding the link to some existing function in other framework...

File: onnx/defs/tensor/defs.cc [hunk 11]
```diff
@@ -2226,6 +2226,65 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            "T",
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:2267]: IIUC, this is used as an index into the input coordinates. Not sure we need to use the same type "T" as the other "T" ... would it make sense to have this be a "T2"?

File: docs/Changelog.md [hunk 12]
```diff
@@ -18939,6 +18939,49 @@ This version of the operator has been available since version 14 of the default
...
+<dd>Three interpolation modes: bilinear (default), nearest and bicubic.</dd>
+<dt><tt>padding_mode</tt> : string (default is zeros)</dt>
+<dd>Supported Padding modes for outside grid values: `zeros`(default), `border`, `reflection`</dd>
+</dl>
+
+#### Inputs
+
+<dl>
+<dt><tt>X</tt> (differentiable) : T</dt>
+<dd>4-D tensor of shape (N, C, inH, inW), where N is the batch size, C is the numbers of channels, inH and inW are the height and width of the data.</dd>
```
Comment (jcwchen) [inline docs/Changelog.md:18966]: nit: Existing op like RoiAlign uses `(N, C, H, W)` to represent the 4d input. Sync the same naming convention might be better.

File: onnx/defs/tensor/defs.cc (hunk 10 above)
Comment (jcwchen) [reply onnx/defs/tensor/defs.cc:2266]: nit: where outH and outW `are` the height and width of offset and output

File: docs/Changelog.md (hunk 3 above)
Comment (hwangdeyu) [reply docs/Changelog.md]: Thanks, added.

File: onnx/defs/tensor/defs.cc (hunk 6 above)
Comment (hwangdeyu) [reply onnx/defs/tensor/defs.cc]: > It is unclear to me what this function is supposed to compute. Can we expand the description to be more complete, or add a link to some existing function in other frameworks, or both? Thanks.

I only find this op in PyTorch, have posed the link in description. And this op is from Spatial Transformer Networks, I also added the paper link too.

File: onnx/defs/tensor/defs.cc (hunk 7 above)
Comment (hwangdeyu) [reply onnx/defs/tensor/defs.cc]: > Please take a look at the example here:
> 
> https://github.com/onnx/onnx/blob/ea8e57a71e01d1e70b188465c028eb3c6c6cc118/onnx/defs/object_detection/defs.cc#L114
> 
> ... `unifyInputDim` should help simplify this.

Thanks, that's very helpful.

File: onnx/defs/tensor/defs.cc (hunk 10 above)
Comment (hwangdeyu) [reply onnx/defs/tensor/defs.cc]: > Is the "offset and output" a typo? Don't understand it.

Added more description..

File: docs/Changelog.md (hunk 12 above)
Comment (hwangdeyu) [reply docs/Changelog.md]: Thanks. Good point.

File: docs/TestCoverage.md (hunk 4 above)
Comment (hwangdeyu) [reply docs/TestCoverage.md:14676]: Actually, I don't find any implementation in numpy for grid_sampler to add the backend tests. Do you have any suggestions to do it?

File: docs/Changelog.md (hunk 2 above)
Comment (hwangdeyu) [reply docs/Changelog.md]: Yes, I think that means a grid generated by affine_grid() should be passed to grid_sample() with the same setting for this option.
refer link: https://pytorch.org/docs/master/generated/torch.nn.functional.affine_grid.html?highlight=align_corners.

File: docs/Changelog.md (hunk 8 above)
Comment (askhade) [reply docs/Changelog.md]: There are couple of issues opened for this op. We can ask there if people need 5-D support if so we can add it now otherwise it is OK to add it later as need arises.

Comment (hwangdeyu) [reply docs/Changelog.md]: According to the example of several github issues that are provided input, all the inputs are 4-D shape. I think we can support 4-D it firstly.
https://github.com/pytorch/pytorch/issues/41957
https://github.com/pytorch/pytorch/issues/27212
https://github.com/AliaksandrSiarohin/first-order-model/issues/92
https://github.com/PaddlePaddle/Paddle2ONNX/issues/273

File: onnx/backend/test/case/node/gridsampler.py [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+
+import onnx
+from ..base import Base
+from . import expect
+
+
+class GridSampler(Base):
+
+    @staticmethod
+    def export_gridsampler():  # type: () -> None
```
Comment (hwangdeyu) [inline onnx/backend/test/case/node/gridsampler.py:18]: Consider using script to generate the test outputs.

Comment (hwangdeyu) [reply onnx/backend/test/case/node/gridsampler.py:18]: added the torch test and hardcoding outputs.

**Comment by gramalingam** [comment]
@hwangdeyu : can  you please look at this comment https://github.com/onnx/sigs/blob/master/operators/meetings/023-20210708.md#issue-treatment-of-reflect-attribute-in-pad-op : it is about the treatment of reflection when padding. It would be helpful to clarify the treatment in this op as well.

**Comment by hwangdeyu** [comment]
> @hwangdeyu : can you please look at this comment https://github.com/onnx/sigs/blob/master/operators/meetings/023-20210708.md#issue-treatment-of-reflect-attribute-in-pad-op : it is about the treatment of reflection when padding. It would be helpful to clarify the treatment in this op as well.

Sure, will add the description of the `reflect` attribute in the `Pad` op. Thanks for reminding.

**Comment by hwangdeyu** [comment]
@gramalingam @askhade @jcwchen Any more comments？Thanks

File: onnx/defs/tensor/defs.cc [hunk 14]
```diff
@@ -2471,6 +2471,110 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            std::string("bilinear"))
+        .Attr(
+            "padding_mode",
+            "Support padding modes for outside grid values: `zeros`(default), `border`, `reflection`. "
+            "zeros: use 0 for out-of-bound grid locations, "
+            "border: use border values for out-of-bound grid locations, "
+            "reflection: use values at locations reflected by the border for out-of-bound grid locations. "
+            "If index 0 represents the margin pixel, the reflected value at index -1 will be the same as the value at index 1. "
+            "For location far away from the border, it will keep being reflected until becoming in bound. "
+            "If pixel location x = -3.5 reflects by border -1 and becomes x' = 1.5, then reflects by border 1 and becomes x'' = -0.5.",
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:2502]: I don't understand this. If x' = 1.5 is reflected in border 1, doesn't that give us x'' = 0.5? Why is is -0.5?

Comment (hwangdeyu) [reply onnx/defs/tensor/defs.cc:2502]: You are right. It also looks like an error in [grid_sample PyTorch documentation](https://pytorch.org/docs/master/generated/torch.nn.functional.grid_sample.html#torch-nn-functional-grid-sample). Will fix this fault.

Comment (hwangdeyu) [reply onnx/defs/tensor/defs.cc]: Fixed, and also post an issue to PyTorch repo. https://github.com/pytorch/pytorch/issues/63559

**Comment by hwangdeyu** [comment]
@gramalingam @askhade 
https://github.com/microsoft/onnxruntime/pull/8551
The `GridSample` operator kernal inplementation is merged into onnxruntime repo as an contrib op.
Any further advise or comment for this ONNX spec PR?

**Review by gramalingam — APPROVED** [review]
LGTM, thanks for the contribution!

**Comment by hwangdeyu** [comment]
@askhade Could we merge this PR now？

**Comment by msollami-sf** [comment]
Someone please merge this

**Comment by askhade** [comment]
@hwangdeyu : This is a new operator and before we merge this we should make sure the spec is finalized. I recommend to wait till the evaluation of the model with this op is complete. I have marked this PR with 1.11 milestone. Next ONNX release will be in ~Nov  so we have some time. Can you work with 1P team to make sure there is no more change required to the spec?

**Comment by hwangdeyu** [comment]
> @hwangdeyu Deyu Huang FTE : This is a new operator and before we merge this we should make sure the spec is finalized. I recommend to wait till the evaluation of the model with this op is complete. I have marked this PR with 1.11 milestone. Next ONNX release will be in ~Nov so we have some time. Can you work with 1P team to make sure there is no more change required to the spec?

Get it. Sure, will connect with them when the evaluation of the model almost finished.

File: onnx/defs/tensor/defs.cc [hunk 15]
```diff
@@ -2520,6 +2520,110 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+For each output location `output[N, C, H_out, W_out]`, the size-2 vector `grid[N, H_out, W_out]` specifies `input` pixel locations `x` and `y`,
+which are used to interpolate the output value `output[N, C, H_out, W_out]`.
+
+The GridSample operator is often used in doing grid generator and sampler in the [Spatial Transformer Networks](https://arxiv.org/abs/1506.02025).
+See also in [torch.nn.functional.grid_sample](https://pytorch.org/docs/master/generated/torch.nn.functional.grid_sample.html#torch-nn-functional-grid-sample).
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    GridSample,
+    15,
```
Comment (askhade) [inline onnx/defs/tensor/defs.cc:2536]: Please change to operator version 16 as this is the current working version.

Comment (hwangdeyu) [reply onnx/defs/tensor/defs.cc]: Done~

Comment (hwangdeyu) [reply onnx/defs/tensor/defs.cc]: > Please change to operator version 16 as this is the current working version. Make sure you regenerate the tests after you update the version.

Anything else required to merge this PR?

**Comment by hwangdeyu** [comment]
@askhade  @gramalingam  Is it a good time to merge this PR？

**Review by askhade — APPROVED** [review]

**Comment by askhade** [comment]
> @askhade @gramalingam Is it a good time to merge this PR？

@hwangdeyu:  Thanks for your contribution! The PR looks good now. I have set it to auto merge. Once all the CIs are successful it will go through.

**Comment by laisimiao** [comment]
> @hwangdeyu : This is a new operator and before we merge this we should make sure the spec is finalized. I recommend to wait till the evaluation of the model with this op is complete. I have marked this PR with 1.11 milestone. Next ONNX release will be in ~Nov so we have some time. Can you work with 1P team to make sure there is no more change required to the spec?

It's Dec. When will onnx 1.11 release?

**Comment by jcwchen** [comment]
Hi @laisimiao,
Because ONNX has released 3 times last year and at that time there weren't enough updates, eventually ONNX decided to postpone 1.11 release. It will probably happen this month or next month. Sorry for the confusion and inconvenience. If you are urgent to use new features after ONNX 1.10, you can try onnx-weekly [package](https://test.pypi.org/project/onnx-weekly/) from TestPyPI. Thank you!
